File: golang/sort_algorithm/bucket_sort.go
package sort_algorithm

import (
	"fmt"
	"math"
	_sort "sort"
)

type BucketSort []int

func (b *BucketSort) SortImpl() {
	defer fmt.Println(b)
	if len(*b) < 2 {
		return
	}
	min, max := (*b)[0], (*b)[0]
	for _, bb := range *b {
		if bb < min {
			min = bb
		}
		if bb > max {
			max = bb
		}
	}
	const bucketNum = 10
	ranges := int(math.Ceil(float64(max-min+1) / float64(bucketNum)))
	bucketList := make([][]int, bucketNum)
	for _, bb := range *b {
		index := (bb - min) / ranges
		bucketList[index] = append(bucketList[index], bb)
	}
	for i := range bucketList {
		_sort.Ints(bucketList[i])
	}
	*b = []int{}
	for i := 0; i < bucketNum; i++ {
		*b = append(*b, bucketList[i]...)
	}
}

File: java/sort_algorithm/impl/BucketSort.java
package sort_algorithm.impl;

import sort_algorithm.Sort;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

public class BucketSort implements Sort {
    @Override
    public void sort(int[] arr) {
        int min = Arrays.stream(arr).min().getAsInt(), max = IntStream.of(arr).max().getAsInt();
        final int bucketNum = 10;
        double range = (max - min + 1)*1.0 / bucketNum;
        List<Integer>[] buckets = new List[bucketNum];
        for (int i = 0; i < buckets.length; i++) {
            buckets[i] = new ArrayList<>();
        }
        for (int i : arr) {
            int index = (int)((i - min) / range);
            buckets[index].add(i);
        }
        for (List<Integer> bucket : buckets) {
            bucket.sort(Integer::compareTo);
        }
        for (int i = 0, j = 0; i < buckets.length; i++) {
            for (Integer integer : buckets[i]) {
                arr[j++] = integer;
            }
        }

        System.out.println(Arrays.toString(arr));
    }
}
